packages/core/src/lib/stream_manager.ts

```diff
@@ -14,60 +18,108 @@ export class StreamManager {
     public addEventListener: Libp2p["addEventListener"]
   ) {
     this.log = new Logger(`stream-manager:${multicodec}`);
-    this.addEventListener(
-      "peer:update",
-      this.handlePeerUpdateStreamPool.bind(this)
-    );
-    this.getStream = this.getStream.bind(this);
     this.streamPool = new Map();
+
+    this.addEventListener("peer:update", this.handlePeerUpdateStreamPool);
+    this.getStream = this.getStream.bind(this);
   }
 
-  public async getStream(peer: Peer): Promise<Stream> {
+  public async getStream(peer: Peer): Promise<Stream | null> {
     const peerIdStr = peer.id.toString();
     const streamPromise = this.streamPool.get(peerIdStr);
 
     if (!streamPromise) {
-      return this.newStream(peer); // fallback by creating a new stream on the spot
+      return this.createStream(peer);
     }
 
-    // We have the stream, let's remove it from the map
     this.streamPool.delete(peerIdStr);
-
     this.prepareNewStream(peer);
 
-    const stream = await streamPromise;
-
-    if (!stream || stream.status === "closed") {
-      return this.newStream(peer); // fallback by creating a new stream on the spot
+    try {
+      const stream = await streamPromise;
+      if (stream && stream.status !== "closed") {
+        return stream;
+      }
+    } catch (error) {
+      this.log.error(`Failed to get stream for ${peerIdStr} -- `, error);
     }
 
-    return stream;
+    return this.createStream(peer);
+  }
+
+  private async createStream(peer: Peer): Promise<Stream | null> {
+    try {
+      return await this.newStream(peer);
+    } catch (error) {
+      this.log.error(
+        `Failed to create a new stream for ${peer.id.toString()} -- `,
+        error
+      );
+      return null;
+    }
   }
 
-  private async newStream(peer: Peer): Promise<Stream> {
+  private async newStream(peer: Peer, retries = 0): Promise<Stream> {
     const connections = this.getConnections(peer.id);
     const connection = selectConnection(connections);
+
     if (!connection) {
       throw new Error("Failed to get a connection to the peer");
     }
-    return connection.newStream(this.multicodec);
+
+    try {
+      return await connection.newStream(this.multicodec);
```

Comment:
let's rewrite to

```ts
let result;
try {
  result = await ...;
}
return result;
```

the reason is due to huge difference in code execution because of `return` or `return await`

Reply:
same for other places

Reply:
can you elaborate on the difference in code execution? imo both of them are the same?

Reply:
if `try { return promise } catch ...` - it won't catch as promise is not awaited
if `try { return await promise } catch ...` - will catch and the difference is only in one word

I think it is better to be on a clear side of things and write:
```
let result;
try {
  result = await ...;
}
return result;
```
as the absence of `await` changes so much (and I faced some nasty bugs in the past because of that)